### PR TITLE
Format `PatternMatchClass`

### LINKED ISSUE #5834: Formatter: `Match` (includes `MatchCase` and `Patterns`)
This is a bigger project that includes the following nodes:

- [x] #6298
- [x] #6299
- [x] #6641
- [x] #6642
- [x] #6644
- [x] #6643
- [x] #6645
- [x] #6557
- [x] #6558
- [x] #6555
- [x] #6753
- [ ] #6933

The match arms have the same considerations for comments as if branches or try-except branches.

### LINKED ISSUE #6642: Format `PatternMatchClass`
Implement formatting for class match patterns

```python
match x:
    case Point2D(0, 0):
        ...

    case Point3D(x=0, y=0, z=0):
        ...
```

### PR DESCRIPTION
**Pull request by LaBatata101**
## Summary
Closes #6642

## Test Plan

`cargo test`

### CODE REVIEW

File: crates/ruff_python_formatter/tests/snapshots/black_compatibility@py_310__pattern_matching_extras.py.snap [hunk 1]
```diff
@@ -248,19 +204,16 @@ match bar1:
...
- 
+@@ -114,6 +125,9 @@
  
  match bar1:
--    case Foo(
+     case Foo(
 -        normal=x, perhaps=[list, {"x": d, "y": 1.0}] as y, otherwise=something, q=t as u
--    ):
-+    case NOT_YET_IMPLEMENTED_PatternMatchClass(0, 0):
++        normal=x,
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/tests/snapshots/black_compatibility@py_310__pattern_matching_extras.py.snap:212]: You'll need to wrap the properties list in an extra group to get this formatting

**Comment by LaBatata101** [comment]
I don't know how to solve the conflict for the `crates/ruff_python_formatter/tests/snapshots/black_compatibility@py_310__pattern_matching_extras.py.snap` file

**Review by MichaReiser — COMMENTED** [review]
So many removed lines ☺️

File: crates/ruff_python_formatter/src/pattern/pattern_match_class.rs [hunk 2]
```diff
@@ -1,19 +1,41 @@
...
+                    join.entry(
+                        key,
+                        &format_with(|f| write!(f, [key.format(), text("="), value.format()])),
+                    );
+                }
+            }
+            join.finish()
+        });
+
+        write!(f, [cls.format(), parenthesized("(", &items, ")")])
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/pattern/pattern_match_class.rs:39]: Try adding an extra group around items

Comment (LaBatata101) [reply crates/ruff_python_formatter/src/pattern/pattern_match_class.rs:39]: Would the function `group` work here?

Comment (MichaReiser) [reply crates/ruff_python_formatter/src/pattern/pattern_match_class.rs:39]: Yes, group(&items) should do the trick if I'm not mistaken

Comment (LaBatata101) [reply crates/ruff_python_formatter/src/pattern/pattern_match_class.rs:39]: Nothing changed for this `crates/ruff_python_formatter/tests/snapshots/black_compatibility@py_310__pattern_matching_extras.py.snap` file when using `group`

File: crates/ruff_python_formatter/tests/snapshots/black_compatibility@py_310__pattern_matching_extras.py.snap [hunk 3]
```diff
@@ -248,19 +204,16 @@ match bar1:
...
++        perhaps=[list, {"NOT_YET_IMPLEMENTED_PatternMatchMapping": _, 2: _}] as y,
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/tests/snapshots/black_compatibility@py_310__pattern_matching_extras.py.snap:213]: Okay, the problem here is that `PatternMatchMapping` isn't yet implemented and the text emitted by the "fake" implementation is much longer than the expected string.

We can ignore this.

File: crates/ruff_python_formatter/src/pattern/pattern_match_class.rs (hunk 2 above)
Comment (MichaReiser) [reply crates/ruff_python_formatter/src/pattern/pattern_match_class.rs:39]: See my comment below. This is fine, assuming we're talking about the same mismatch.

**Review by MichaReiser — COMMENTED** [review]
This looks good to me. Thank you. 

Could you add one test case to 

https://github.com/astral-sh/ruff/blob/main/crates/ruff_python_formatter/resources/test/fixtures/ruff/statement/match.py#L285


that tests for the following weird comment placement

```python
match x:
	case (
		Point2D
		# awkward
		(x = 2)
	: ...
```
```

File: crates/ruff_python_formatter/src/pattern/pattern_match_class.rs [hunk 4]
```diff
@@ -1,19 +1,41 @@
...
+        } = item;
+
+        let items = format_with(|f| {
+            let mut join = f.join_comma_separated(range.end());
+
+            if !patterns.is_empty() {
+                join.nodes(patterns.iter());
+            }
+
+            if !kwd_attrs.is_empty() {
```
Comment (konstin) [inline crates/ruff_python_formatter/src/pattern/pattern_match_class.rs:28]: Do we need the `if !kwd_attrs.is_empty()`?

File: crates/ruff_python_formatter/resources/test/fixtures/ruff/statement/match.py [hunk 5]
```diff
@@ -263,3 +263,30 @@ def foo():
...
+    case FooBar(# leading
+# leading
+    # leading
+        # leading
+            0 # trailing
+# trailing
+    # trailing
+        # trailing
+            ):
+        ...
```
Comment (konstin) [inline crates/ruff_python_formatter/resources/test/fixtures/ruff/statement/match.py:292]: Could you add something like
```python
match a:
    case A(
        b # b
        = # c
        2 # d
        # e
    ):
        pass
```
with the left hand side and the right hand side split of the assignment split by comment? I wanted to add an additional comment before the `b` but that's blocked by https://github.com/astral-sh/ruff/issues/6866

**Review by konstin — APPROVED** [review]

**Comment by charliermarsh** [comment]
Will review and merge today. I can also follow-up with new tests once I fix https://github.com/astral-sh/ruff/issues/6866.

**Comment by charliermarsh** [comment]
Okay, I had to make some changes to the PR to adapt to our parenthesis handling for patterns that changed a bit upstream (mostly, to ensure we handle the single-argument case correctly, like we do for call expressions). We also now handle comments before the arguments, like we do for call expressions.

There are several comment cases that we're still getting wrong (valid syntax, but slightly incorrect placement), but that are really hard to fix without changing the AST. (Namely, we need an `Arguments` node, and we need a single node for `x=1` -- we have both of these things for call expressions so we should have them here too.)

I'm going to update the AST, then do another pass on this node.

File: crates/ruff_python_formatter/tests/snapshots/format@statement__match.py.snap [hunk 6]
```diff
@@ -825,6 +884,73 @@ match foo:
...
+        ()
+    ):
+        ...
+
+    case Point2D(  # end of line line
+    ):
+        ...
+
+    case Point2D(
+        # end of line
```
Comment (charliermarsh) [inline crates/ruff_python_formatter/tests/snapshots/format@statement__match.py.snap:907]: This is not quite right (it should be immediately after the `(`, but very hard to fix without a node for the arguments.

**Comment by charliermarsh** [comment]
Filed here: https://github.com/astral-sh/ruff/issues/6880. I'm on it.

**Pull request merged**